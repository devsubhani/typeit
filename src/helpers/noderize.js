export default function (string) {

  let parser = new DOMParser();
  let doc = parser.parseFromString(string, "text/html");
  let nodes = [].slice.call(doc.body.querySelectorAll('*'));

  //-- Replace node instances with placeholders.
  nodes.forEach(item => {
    string = string.replace(item.outerHTML, '{%}');
  });

  let stringArray = string.split('');

  //-- Replace placeholders w/ nodes.
  stringArray.forEach((item, index) => {

    //-- Check for a placeholder.
    if (
      item === '{' &&
      stringArray[index + 1] === '%' &&
      stringArray[index + 2] === '}'
    ) {
      //-- Insert element.
      stringArray.splice(index, 3, nodes.shift());
    }
  });

  return stringArray;
}
