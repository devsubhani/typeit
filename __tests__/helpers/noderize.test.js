import noderize from '../../src/helpers/noderize';

test("Parses normal string correctly.", () => {

  let result = noderize(
    'Hello, this is my string.'
  );

  expect(result).toMatchSnapshot();
});

test("Parses single HTML tag.", () => {

  let result = noderize(
    'Hello, this is some <strong>bold</strong> text.'
  );

  expect(result).toMatchSnapshot();
});

test("Parses multiple HTML tags.", () => {

  let result = noderize(
    'Hello, this is some <strong>bold</strong> text, and some <i>italicized</i> text.'
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at beginning of string.", () => {

  let result = noderize(
    '<strong>Hello!</strong> This is some text with HTML at the beginning.'
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at end of string.", () => {

  let result = noderize(
    'This is some text with HTML at the <em>beginning.</em>'
  );

  expect(result).toMatchSnapshot();
});
