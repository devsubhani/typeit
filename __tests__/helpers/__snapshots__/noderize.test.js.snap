// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses HTML tag at beginning of string. 1`] = `
Array [
  Object {
    "attributes": Array [],
    "content": "H",
    "isFirstCharacter": true,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "!",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  " ",
  "T",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  "t",
  "e",
  "x",
  "t",
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "H",
  "T",
  "M",
  "L",
  " ",
  "a",
  "t",
  " ",
  "t",
  "h",
  "e",
  " ",
  "b",
  "e",
  "g",
  "i",
  "n",
  "n",
  "i",
  "n",
  "g",
  ".",
]
`;

exports[`Parses HTML tag at end of string. 1`] = `
Array [
  "T",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  "t",
  "e",
  "x",
  "t",
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "H",
  "T",
  "M",
  "L",
  " ",
  "a",
  "t",
  " ",
  "t",
  "h",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "g",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "g",
    "isFirstCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": ".",
    "isFirstCharacter": false,
    "tag": "EM",
  },
]
`;

exports[`Parses HTML tag with attributes. 1`] = `
Array [
  "T",
  "h",
  "i",
  "s",
  " ",
  "s",
  "t",
  "r",
  "i",
  "n",
  "g",
  " ",
  "h",
  "a",
  "s",
  " ",
  "a",
  "n",
  " ",
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": true,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "m",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "n",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "t",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "a",
  "t",
  "t",
  "r",
  "i",
  "b",
  "u",
  "t",
  "e",
  "s",
  ".",
]
`;

exports[`Parses multiple HTML tags. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ",",
  " ",
  "a",
  "n",
  "d",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": true,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "t",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "a",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "c",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "z",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "tag": "I",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ".",
]
`;

exports[`Parses normal string correctly. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "m",
  "y",
  " ",
  "s",
  "t",
  "r",
  "i",
  "n",
  "g",
  ".",
]
`;

exports[`Parses single HTML tag. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "tag": "STRONG",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ".",
]
`;
